contracts/evmscript/EVMScriptRunner.sol

```diff
@@ -40,33 +39,35 @@ contract EVMScriptRunner is AppStorage, Initializable, EVMScriptRegistryConstant
 
         bytes4 sig = executor.execScript.selector;
         bytes memory data = abi.encodeWithSelector(sig, _script, _input, _blacklist);
-        require(address(executor).delegatecall(data), ERROR_EXECUTION_REVERTED);
+        bool result = address(executor).delegatecall(data);
 
-        bytes memory output = returnedDataDecoded();
-
-        emit ScriptResult(address(executor), _script, _input, output);
-
-        return output;
-    }
-
-    /**
-    * @dev Copies and returns last's call data. Needs to ABI decode first
-    */
-    function returnedDataDecoded() internal pure returns (bytes) {
-        bytes memory ret;
+        bytes memory output;
         assembly {
             let size := returndatasize
 
-            ret := mload(0x40) // free mem ptr get
-            mstore(0x40, add(ret, add(size, 0x20))) // free mem ptr set
+            output := mload(0x40) // free mem ptr get
+            mstore(0x40, add(output, add(size, 0x20))) // free mem ptr set
 
-            switch size
-            case 0 {}
+            // If the call returned error data, forward it
+            switch result
+            case 0 {
+                returndatacopy(output, 0, size) // copy full return data
+                revert(output, size)
+            }
             default {
-                returndatacopy(ret, 0x20, sub(size, 0x20)) // copy return data
+                switch size
+                case 0 {}
+                default {
+                    // Copy and ABI decode result, since `execScript()`'s return type is `bytes`
+                    mstore(output, size)
```

Comment:
I think this is actually something we missed before; we weren't setting the size correctly in the old output.

Reply:
Good catch

Reply:
The first word is used to tell the return data size?

Reply:
See https://github.com/aragon/aragonOS/pull/496/files#r268555232.

Although now I'm thinking; maybe this "ABI decode" is unnecessary and we can just do

```solidity
returndatacopy(output, 0x0, size)
```

Since the bytes length would be the same (as opposed to something like a uint256 array)?

-----------

In any case, this is still wrong; it should be `sub(size, 0x20)`.

Reply:
Exactly, I think we can just do `returndatacopy(output, 0x0, returndatasize)`

Reply:
Recording an offline conversation: because the returned data is ABI encoded, the actual payload looked like this:

```
[ position of the first dynamic length return value = 0x20 (32 bytes) ]
[ output length (32 bytes) ]
[ output content (N bytes) ]
```

Because of this we could copy into the output array all the returned bytes minus the first word (as we assume the return types) directly with: `returndatacopy(ret, 0x20, sub(returndatasize, 0x20))`